### PR TITLE
Indicator

### LINKED ISSUE #53: Implement <Indicator /> component
The blue circle in the top left corner in the picture below, which means an unread message:

Image

Probably relates to:
- [x] #58

### PR DESCRIPTION
**Pull request by vitalii-ilchenko-ssa-group**
Closes #53

### CODE REVIEW

File: packages/core/src/components/Indicator/types.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  isVisible: boolean;
+  position?: 'right' | 'left';
+  background?: string;
+  text?: string | number | JSX.Element;
+  children?: React.ReactNode;
+}
+
+export interface ChildrenDataProps {
+  [key: string]: number;
+}
```
Comment (react-developer-ssa-group) [inline packages/core/src/components/Indicator/types.ts:11]: Could we use `Record<string, number>` here?

File: packages/core/src/components/Indicator/Indicator.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                  border-radius: 3px;
+                  aspect-ratio: 0;
+                `,
+            ]
+          }
+          background={background}>
+          {text}
+        </IndicatorBase>
+      ) : null}
+      <div ref={childrenRef} css={{ width: 'fit-content' }}>
```
[the remark this answers is not in the document]
Comment (arthur-kashaiev-ssa-group) [reply packages/core/src/components/Indicator/Indicator.tsx:67]: Let's attach the ref to the child component. And make it the requirement: the child component should accept a `ref` prop

File: packages/core/src/components/Indicator/types.ts (hunk 1 above)
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/types.ts:11]: Done

File: packages/core/src/components/Indicator/Indicator.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/Indicator.tsx:67]: Done

File: packages/core/src/components/Indicator/Indicator.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    <React.Fragment>
+      {isVisible ? (
+        <IndicatorBase
+          data-testid={`indicator-${position}`}
+          ref={indicatorRef}
+          css={
+            childrenData && [
+              css`
+                top: ${childrenData.top + 4}px;
+                left: ${position === 'right' && '-8px'};
```
[the remark this answers is not in the document]
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/Indicator.tsx:46]: Done

File: packages/core/src/components/Indicator/types.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+export interface IndicatorProps {
+  isVisible: boolean;
+  position?: 'right' | 'left';
+  background?: string;
+  text?: string | number | JSX.Element;
+  children?: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/types.ts:6]: Done

File: packages/core/src/components/Indicator/Indicator.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    getByTestId('indicator-left');
+    getByText('Content');
+    getByText('+20');
+  });
+
+  it('Is not rendered in the document', () => {
+    const { queryByTestId } = render(<Indicator isVisible={false} />);
+
+    expect(queryByTestId('indicator-left')).not.toBeInTheDocument;
```
[the remark this answers is not in the document]
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/Indicator.spec.tsx:44]: Done

File: packages/core/src/components/Indicator/Indicator.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  it('Renders with right position', () => {
+    const { getByTestId } = render(
+      <Indicator isVisible={true} position="right" />,
+    );
+
+    getByTestId('indicator-right');
+  });
+
+  it('Renders with children and default left position', () => {
```
[the remark this answers is not in the document]
Comment (vitalii-ilchenko-ssa-group) [reply packages/core/src/components/Indicator/Indicator.spec.tsx:18]: done

**Review by arthur-kashaiev-ssa-group — APPROVED** [review]

**Review by react-developer-ssa-group — DISMISSED** [review]

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 12       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 2  | ❗       |
    


<details>
<summary>🟡 storybook/components-indicator--default__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--default__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--default__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--default__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--right-side__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--right-side__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--right-side__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--right-side__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



*❗ storybook/components-card--icon-header__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/components-card--item-card-avatar__[w390px]*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 6  | ❗       |
    


<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-meal-nutrients--default__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-progress-info--default__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>

**Review by arthur-kashaiev-ssa-group — DISMISSED** [review]

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 6       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 3  | ❗       |
    


<details>
<summary>🟡 storybook/components-indicator--default__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--right-side__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/components-indicator--with-content__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



*❗ storybook/components-card--icon-header__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/components-card--item-card__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/utility-responsive-image--responsive-image__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 6  | ❗       |
    


<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-heart-rate--hint-opened__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-meal-nutrients--default__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-progress-info--default__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>

File: packages/core/src/components/Indicator/Indicator.tsx [hunk 7]
```diff
@@ -17,7 +18,10 @@ const Indicator = ({
   const [childrenData, setChildrenData] = useState<ChildrenDataProps | null>(
     {} || null,
   );
+  const { width: windowWidth } = useWindowSize();
+  const theme = useTheme();
 
+  console.log();
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Indicator/Indicator.tsx:24]: Remove this `console.log()`

**Review by arthur-kashaiev-ssa-group — CHANGES_REQUESTED** [review]

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 2       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 12  | ❗       |
    


*🟡 storybook/components-indicator--default__[w390px]*: 


| Image |
| --- |
|  |



*🟡 storybook/components-indicator--right-side__[w390px]*: 


| Image |
| --- |
|  |



<details>
<summary>❗ storybook/components-card--item-card__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--default__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--default__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--default__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--right-side__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--right-side__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--right-side__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--with-content__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--with-content__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--with-content__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-indicator--with-content__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/utility-responsive-image--responsive-image__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 4  | ❗       |
    


*❗ custom/widgets-heart-rate--hint-opened__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-heart-rate--hint-opened__[w390px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-heart-rate--hint-opened__[w900px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-heart-rate--default__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |

**Review by arthur-kashaiev-ssa-group — APPROVED** [review]

**Pull request merged**